packages/rum-core/src/performance-monitoring/metrics/inp/process.js

```diff
@@ -0,0 +1,256 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2017-present, Elasticsearch BV
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import { EVENT, FIRST_INPUT } from '../../../common/constants'
+import { isPerfInteractionCountSupported } from '../../../common/utils'
+import { PerfEntryRecorder } from '../metrics'
+
+// Threshold recommended by Google
+const INP_THRESHOLD = 40
+
+// The Google guideline is just to track the 10 longest interactions as INP candidates
+const MAX_INTERACTIONS_TO_CONSIDER = 10
+
+const inpState = {
+  minInteractionId: Infinity,
+  maxInteractionId: 0,
+  interactionCount: 0,
+  longestInteractions: []
+}
+
+/**
+ * Observes Event Timing API entries
+ */
+export function observeUserInteractions(
+  recorder = new PerfEntryRecorder(processUserInteractions)
+) {
+  /**
+   * We set the lowest threshold (16) supported by browsers
+   * this way we can estimate the total number of different user interactions which is essential to calculate P98, among other things.
+   * Once browsers start exposing the `performance.interactionCount` property the
+   * threshold will be 40, which is the value Google recommends as the default threshold
+   * https://github.com/GoogleChrome/web-vitals/blob/3806160ffbc93c3c4abf210a167b81228172b31c/src/onINP.ts#L203
+   */
+  const durationThreshold = isPerfInteractionCountSupported()
```

Comment:
I declared a variable on top of this function. (so we don't call it twice there)

But I would keep this as a util since makes testing more flexible. (at the end of the day we are only invoking it twice anyway)